src/test/regress/regress_gp.c

```diff
@@ -2446,3 +2446,113 @@ gp_get_int_tuples(PG_FUNCTION_ARGS)
 
 	SRF_RETURN_DONE(fctx);
 }
+
+typedef struct
+{
+	int cur_tuple_idx;
+	int cur_segment_idx;
+	int cur_segment_tuple_idx;
+
+	int n_segments;
+	int n_tuples;
+
+	Datum some_text;
+	struct pg_result **pg_results;
+} gp_mock_cdbdispatchcommand_status;
+
+/* 
+ * This test function mocks CdbDispatchCommand() with a customizable amount of
+ * tuples.
+ */
+PG_FUNCTION_INFO_V1(gp_mock_cdbdispatchcommand);
+Datum
+gp_mock_cdbdispatchcommand(PG_FUNCTION_ARGS)
+{
+	FuncCallContext *func_ctx;
+	gp_mock_cdbdispatchcommand_status *my_status;
+
+	int arg_tuple_amount = PG_GETARG_INT32(0);
+
+	if (arg_tuple_amount <= 0)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("gp_mock_cdbdispatchcommand() should only be "
+							   "called with it's parameter greater than 0")));
+	}
+
+	if (SRF_IS_FIRSTCALL())
+	{
+		func_ctx = SRF_FIRSTCALL_INIT();
+
+		MemoryContext oldcontext =
+			MemoryContextSwitchTo(func_ctx->multi_call_memory_ctx);
+
+		my_status = palloc0(sizeof(gp_mock_cdbdispatchcommand_status));
+
+		/* Cache the return result. */
+		my_status->some_text = CStringGetTextDatum("sometext");
+
+		/* Send the command to segments from QD. */
+		if (Gp_role == GP_ROLE_DISPATCH)
+		{
+			char *query =
+				psprintf("SELECT * FROM gp_mock_cdbdispatchcommand(%d)",
+						 arg_tuple_amount);
+
+			CdbPgResults cdb_pgresults = {0};
+			CdbDispatchCommand(query, DF_WITH_SNAPSHOT, &cdb_pgresults);
+
+			pfree(query);
+
+			Assert(cdb_pgresults.numResults > 0);
+
+			for (int i = 0; i < cdb_pgresults.numResults; i++)
+			{
+				Assert(PQresultStatus(cdb_pgresults.pg_results[i]) ==
+					   PGRES_TUPLES_OK);
+				my_status->n_tuples += PQntuples(cdb_pgresults.pg_results[i]);
+			}
+
+			my_status->n_segments = cdb_pgresults.numResults;
+			my_status->pg_results = cdb_pgresults.pg_results;
+		}
+
+		func_ctx->user_fctx = my_status;
+
+		MemoryContextSwitchTo(oldcontext);
+	}
+
+	func_ctx = SRF_PERCALL_SETUP();
+	my_status = func_ctx->user_fctx;
+
+	/* Generate fake tuples from every segment. */
+	if (my_status->cur_tuple_idx++ < arg_tuple_amount)
```

Comment:
Why do we need to increment `my_status->cur_tuple_idx` after `SRF_RETURN_NEXT(func_ctx, ret);` below on the coordinator?